### PR TITLE
cpanm - revamp module

### PR DESCRIPTION
**Pull request by russoz**
##### SUMMARY
This PR uses code written in #1976 but expands on that. That PR lagged too far behind `main`, up to the point where it was too time-consuming to rebase, so we closed that one and brought the changes over to here.

In the wake of #208 the module `cpanm` was rewritten with a number of improvements. 

##### ISSUE TYPE
- Docs Pull Request
- Feature Pull Request

##### COMPONENT NAME
plugins/modules/packaging/language/cpanm.py
plugins/module_utils/module_helper.py

### CODE REVIEW

File: tests/integration/targets/cpanm/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    state: present
+  become: yes
+  when: ansible_os_family == "Debian"
+
+- name: install a Perl package
+  cpanm:
+    name: JSON
+    notest: yes
+  register: install_perl_package_result
+  ignore_errors: yes
```
Comment (aminvakil) [inline tests/integration/targets/cpanm/tasks/main.yml:26]: Is it ok for this task to fail? What is the reason of ignore errors of this task and assert that it's not failed in next task?

File: tests/integration/targets/cpanm/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      - install_perl_package_result is not failed
+
+- name: install a Perl package with version operator
+  cpanm:
+    name: JSON
+    version: "@4.01"
+    notest: yes
+    mode: new
+  register: install_perl_package_with_version_op_result
+  ignore_errors: yes
```
Comment (aminvakil) [inline tests/integration/targets/cpanm/tasks/main.yml:41]: Also here.

**Review by aminvakil — COMMENTED** [review]
Thank you for this follow-up PR, I know this is still WIP, but I was curious about `ignore_errors` :)

File: tests/integration/targets/cpanm/tasks/main.yml (hunk 1 above)
Comment (russoz) [reply tests/integration/targets/cpanm/tasks/main.yml:26]: Technically I could remove that line off this one play, but the reason for doing that here is that we want it to fail(or not) when doing the assert (below).

Comment (russoz) [reply tests/integration/targets/cpanm/tasks/main.yml:26]: This is mandatory when testing things that *must* fail.

Comment (russoz) [reply tests/integration/targets/cpanm/tasks/main.yml:26]:
```suggestion
```

File: tests/integration/targets/cpanm/tasks/main.yml (hunk 2 above)
Comment (russoz) [reply tests/integration/targets/cpanm/tasks/main.yml:40]:
```suggestion
```

**Comment by russoz** [comment]
@aminvakil could you please take another look at this one? TIA
@felixfontein I thought I wouldn't be able to make it, but I managed to, there it goes. If there's still time for it to make 2.5.0 :-)

**Comment by aminvakil** [comment]
I looked over integration tests and added documentation about module, and it LGTM.

File: plugins/modules/packaging/language/cpanm.py [hunk 3]
```diff
@@ -49,22 +51,53 @@
...
+    description:
+      - Controls the module behavior. See notes below for more details.
+    type: str
+    choices: [compatibility, new]
+    default: compatibility
+  name_check:
+    description:
+      - When in C(new) mode, this parameter can be used to check if there is a module I(name) installed (at I(version), when specified).
+    type: str
+    version_added: 2.5.0
```
Comment (felixfontein) [inline plugins/modules/packaging/language/cpanm.py:80]:
```suggestion
    version_added: 3.0.0
```

**Comment by felixfontein** [comment]
@russoz sorry I wasn't able to get it reviewed before 2.5.0.

**Comment by russoz** [comment]
No rush :-) Thanks for doing that now.

File: plugins/modules/packaging/language/cpanm.py [hunk 4]
```diff
@@ -13,58 +14,90 @@
...
   installdeps:
     description:
-      - Only install dependencies
+      - Only install dependencies.
     type: bool
     default: no
   version:
     description:
-      - minimum version of perl module to consider acceptable
+      - Version specification for the perl module. When I(mode) is (C(new), C(cpanm) version operators are accepted.
```
Comment (felixfontein) [inline plugins/modules/packaging/language/cpanm.py:54]:
```suggestion
      - Version specification for the perl module. When I(mode) is C(new), C(cpanm) version operators are accepted.
```

File: plugins/modules/packaging/language/cpanm.py [hunk 5]
```diff
@@ -13,58 +14,90 @@
...
     description:
-      - Override the path to the cpanm executable
+      - Override the path to the cpanm executable.
     type: path
+  mode:
+    description:
+      - Controls the module behavior. See notes below for more details.
+    type: str
+    choices: [compatibility, new]
+    default: compatibility
```
Comment (felixfontein) [inline plugins/modules/packaging/language/cpanm.py:75]:
```suggestion
    default: compatibility
    version_added: 3.0.0
```

File: plugins/modules/packaging/language/cpanm.py [hunk 6]
```diff
@@ -109,105 +142,114 @@
...
+        mirror=dict(fmt=('--mirror', '{0}'),),
+        mirror_only=dict(fmt="--mirror-only", style=ArgFormat.BOOLEAN),
+        installdeps=dict(fmt="--installdeps", style=ArgFormat.BOOLEAN),
+        system_lib=dict(fmt="--sudo", style=ArgFormat.BOOLEAN),
     )
+    check_rc = True
+
+    def __init_module__(self):
+        if self.vars.mode == "compatibility" and self.vars.name_check:
+            raise ModuleHelperException("Parameter name_check can only be used with mode=new")
```
Comment (felixfontein) [inline plugins/modules/packaging/language/cpanm.py:184]: Is both this...

File: plugins/modules/packaging/language/cpanm.py [hunk 7]
```diff
@@ -109,105 +142,114 @@
...
+        v = self.vars
+
+        self.command = self.module.get_bin_path(v.executable if v.executable else self.command)
+        self.vars.set("binary", self.command)
+
+        pkg_param = 'from_path' if v.from_path else 'name'
+
+        if v.mode == 'compatibility':
+            if v.name_check:
+                raise ModuleHelperException("Parameter 'name_check' can only be used with 'mode=new'")
```
Comment (felixfontein) [inline plugins/modules/packaging/language/cpanm.py:222]: ...and this needed? I guess one of them is enough.

File: plugins/modules/packaging/language/cpanm.py [hunk 8]
```diff
@@ -109,105 +142,114 @@
...
+    def __init_module__(self):
+        if self.vars.mode == "compatibility" and self.vars.name_check:
+            raise ModuleHelperException("Parameter name_check can only be used with mode=new")
+
+    def _is_package_installed(self, name, locallib, version):
+        if name is None or name.endswith('.tar.gz'):
+            return False
+
+        env = {"PERL5LIB": "%s/lib/perl5" % locallib} if locallib else {}
+        cmd = ['perl', '-le', 'use %s %s;' % (name, version)]
```
Comment (felixfontein) [inline plugins/modules/packaging/language/cpanm.py:191]: Shouldn't this handle the case `version is None` as well?

Comment (russoz) [reply plugins/modules/packaging/language/cpanm.py]: I think I was shuffling these things around and forgot that I left two checks in place. Thanks for spotting.

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
@russoz thanks a lot for working on this!
@BerkhanBerkdemir thanks for initiating this!
@aminvakil thanks for reviewing!

**Pull request merged**